// ionic-long-press/bundles/ionic-long-press.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('ionic-angular/gestures/gesture'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ionic-long-press', ['exports', '@angular/core', 'ionic-angular/gestures/gesture', '@angular/common'], factory) :
    (factory((global['ionic-long-press'] = {}),global.ng.core,null,global.ng.common));
}(this, (function (exports,core,gesture,common) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var LongPressDirective = /** @class */ (function () {
        function LongPressDirective(zone, el) {
            this.zone = zone;
            this.onPressStart = new core.EventEmitter();
            this.onPressing = new core.EventEmitter();
            this.onPressEnd = new core.EventEmitter();
            this.el = el.nativeElement;
        }
        /**
         * @return {?}
         */
        LongPressDirective.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                var _this = this;
                if (!this.interval)
                    this.interval = 500;
                if (this.interval < 40) {
                    throw new Error('A limit of 40ms is imposed so you don\'t destroy device performance. If you need less than a 40ms interval, please file an issue explaining your use case.');
                }
                this.pressGesture = new gesture.Gesture(this.el);
                this.pressGesture.listen();
                this.pressGesture.on('press', function (e) {
                    _this.onPressStart.emit(e);
                    _this.clearInt();
                    _this.int = /** @type {?} */ (setInterval(function () {
                        _this.onPressing.emit();
                    }, _this.interval));
                });
                this.pressGesture.on('pressup', function (e) {
                    _this.pressEnd();
                });
                this.pressGesture.on('pan', function (e) {
                    _this.pressEnd();
                });
                this.pressGesture.on('release', function (e) {
                    _this.pressEnd();
                });
                this.el.addEventListener('mouseleave', function (e) {
                    _this.pressEnd();
                });
                this.el.addEventListener('mouseout', function (e) {
                    _this.pressEnd();
                });
            };
        /**
         * @return {?}
         */
        LongPressDirective.prototype.clearInt = /**
         * @return {?}
         */
            function () {
                if (this.int !== undefined) {
                    clearInterval(this.int);
                    this.int = undefined;
                }
            };
        /**
         * @return {?}
         */
        LongPressDirective.prototype.pressEnd = /**
         * @return {?}
         */
            function () {
                this.clearInt();
                this.onPressEnd.emit();
            };
        /**
         * @return {?}
         */
        LongPressDirective.prototype.ngOnDestroy = /**
         * @return {?}
         */
            function () {
                this.pressEnd();
                this.pressGesture.destroy();
            };
        LongPressDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[ion-long-press]'
                    },] },
        ];
        /** @nocollapse */
        LongPressDirective.ctorParameters = function () {
            return [
                { type: core.NgZone, },
                { type: core.ElementRef, },
            ];
        };
        LongPressDirective.propDecorators = {
            "interval": [{ type: core.Input },],
            "onPressStart": [{ type: core.Output },],
            "onPressing": [{ type: core.Output },],
            "onPressEnd": [{ type: core.Output },],
        };
        return LongPressDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var LongPressModule = /** @class */ (function () {
        function LongPressModule() {
        }
        LongPressModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [
                            common.CommonModule
                        ],
                        declarations: [LongPressDirective],
                        exports: [LongPressDirective]
                    },] },
        ];
        return LongPressModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    exports.LongPressDirective = LongPressDirective;
    exports.LongPressModule = LongPressModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
